File: internal/secrets/seal/sas.go
package seal

import (
	"crypto/sha256"
	"fmt"
)

const sasMagic = "sas/v1"

// DeriveSAS computes the enrollment Short Authentication String (SAS) from the
// chain state and the new device's public keys.
//
// Both the approver and the enrollee independently compute this code; a human
// confirms the two codes match out-of-band before the approver commits the
// add-entry (spec §2 [WM4]). The code is NEVER parsed from the enrollment
// link — a link-carried code authenticates nothing.
//
// Formula:
//
//	SHA-256(encodeFields("sas/v1", genesis_hash, head_hash, new_x25519_pub, new_sign_pub))
//
// Output: first 6 bytes of the digest (48 bits) encoded as base-36
// (lowercase alphanum), chunked as "xxxx-xxxx-xxxx" for readability.
//
// Cross-language note: must match web/src/keys/sas.ts:deriveSAS exactly.
// Both encodeFields implementations use big-endian uint64 length prefixes per
// field; the base-36 encoding takes first 6 bytes MSB-first and iterates 12
// positions right-to-left, filling with '0' on leading zeros.
func DeriveSAS(genesisHash, headHash, newX25519Pub, newSignPub []byte) (string, error) {
	if len(genesisHash) == 0 || len(headHash) == 0 {
		return "", fmt.Errorf("seal: DeriveSAS: genesis_hash and head_hash must be non-empty")
	}
	digest := sha256.Sum256(encodeFields(
		[]byte(sasMagic),
		genesisHash,
		headHash,
		newX25519Pub,
		newSignPub,
	))
	// Extract first 6 bytes (48 bits) as a big-endian uint64 (top 16 bits unused).
	n := uint64(digest[0])<<40 | uint64(digest[1])<<32 | uint64(digest[2])<<24 |
		uint64(digest[3])<<16 | uint64(digest[4])<<8 | uint64(digest[5])
	const base = uint64(36)
	const chars = "0123456789abcdefghijklmnopqrstuvwxyz"
	code := make([]byte, 12)
	for i := 11; i >= 0; i-- {
		code[i] = chars[n%base]
		n /= base
	}
	return fmt.Sprintf("%s-%s-%s", code[0:4], code[4:8], code[8:12]), nil
}

File: web/src/keys/sas.ts
/**
 * SAS (Short Authentication String) derivation for enrollment (Phase 5, [WM4]).
 *
 * Both sides independently derive the verification code from:
 *   SHA-256(encodeFields("sas/v1", genesis_hash, head_hash, new_x25519_pub, new_sign_pub))
 *
 * The code is NEVER parsed from the enrollment link — a code carried in the
 * link authenticates nothing ([WM4]). Both sides compute it independently
 * from their own view of the data; the approver from the pubkeys it received
 * in the link, the enrollee from its own generated keys.
 *
 * Encoding: first 6 bytes of the SHA-256 output encoded as base-36 (lowercase
 * alphanum), chunked into 3×4 characters for readability.
 * Entropy: 48 bits, exceeds the ≥48-bit spec requirement.
 */

import { encodeFields, sha256 } from "./encoding";

const SAS_VERSION = new TextEncoder().encode("sas/v1");

/**
 * deriveSAS computes the enrollment SAS code from the chain anchors and the
 * new device's public keys.
 *
 * Returns a 12-character human-comparable string (3 groups of 4, e.g. "a4bx-r2zy-8t3k").
 */
export async function deriveSAS(
  genesisHash: Uint8Array,
  headHash: Uint8Array,
  newX25519Pub: Uint8Array,
  newSignPub: Uint8Array,
): Promise<string> {
  const digest = await sha256(
    encodeFields(SAS_VERSION, genesisHash, headHash, newX25519Pub, newSignPub),
  );
  // Take first 6 bytes (48 bits) and encode as base-36
  const n = (
    BigInt(digest[0]) << 40n |
    BigInt(digest[1]) << 32n |
    BigInt(digest[2]) << 24n |
    BigInt(digest[3]) << 16n |
    BigInt(digest[4]) << 8n |
    BigInt(digest[5])
  );
  // Encode as at least 10-char base-36 string, left-padded with 'a' (char 10)
  const BASE = 36n;
  const chars = "0123456789abcdefghijklmnopqrstuvwxyz";
  let val = n;
  let code = "";
  for (let i = 0; i < 12; i++) {
    code = chars[Number(val % BASE)] + code;
    val /= BASE;
  }
  // Split into 3 groups of 4
  return `${code.slice(0, 4)}-${code.slice(4, 8)}-${code.slice(8, 12)}`;
}

/**
 * verifyEnrollmentSAS checks that the SAS the approver computed matches the
 * SAS the user compared out-of-band. Constant-time comparison is not required
 * here (the SAS is displayed to both humans — timing is irrelevant), but we
 * keep a strict equality check.
 */
export function verifySAS(computed: string, displayed: string): boolean {
  return computed === displayed;
}
